### ISSUE TITLE
Removing team member while Poker meeting is open removes all meeting contents

### ISSUE THREAD

**Issue by enriquesanchez** [ParabolInc/parabol#7625]
[From a customer email](https://parabol.slack.com/archives/C4JAUUZ9P/p1672868782689139) 🔒, steps to reproduce:

1. Create Poker meeting on a team with at least 2 members
2. Have team members interact with the meeting (add issues, vote, no-vote, abstain, etc)
3. While the meeting is open, go to the team settings and remove one of the members from the team who interacted with the meeting
4. Open the meeting again and you'll notice all contents are gone: all issues that had been added, and the votes that had already been submitted, will be lost. The meeting is stuck in the Scope phase and it’s labeled as completed. The only action that the Facilitator can take at this point is to end the meeting. The Next button is gone.

## Acceptance Criteria
When a team member who is participating in an active Poker meeting is removed from the team:
   - No contents from said meeting are lost
   - The facilitator and other team members can still interact with the meeting

Triage is performed to find the root cause of the bug, timeboxed to ~1 hour. 

## Estimated effort
1 hour to triage. More if root cause is already identified.

**Comment by jordanh** [comment]
This bug was opened in response to a email sent from me to a customer who had been using Parabol and stopped. My email asks them why they stopped. They told me it was because they encountered this bug.

I asked @enriquesanchez if he could reproduce it and if so, to please open a bug. He was able to reproduce it. I will prioritized it accordingly

**Comment by jmtaber129** [comment]
@enriquesanchez in your repro, are the stages + votes that are lost limited to those submitted by the removed user, or are _all_ stages + votes lost? I just tried to reproduce, and only the issues added by the removed user and the votes submitted by the removed user were lost. Both of these seem to be intentionally removed in the code.

**Comment by enriquesanchez** [comment]
@jmtaber129 I just tried this again, and you are correct, the issues that were added by the removed user are the only ones that were lost.

I think I got this part wrong on my initial repro attempt:

>The meeting is stuck in the Scope phase and it’s labeled as completed. The only action that the Facilitator can take at this point is to end the meeting. The Next button is gone.

The meeting is in this state because all the issues (which happened to be added by the removed user when I reproed it) are gone. But I can still add new ones and then move the meeting forward.

That said, I tested it three times, and on the second and third attempt I as the facilitator was not able to easily end the meeting. Clicking on `End meeting` did nothing. I did get the summary email though, and it was until I refreshed the page that the meeting was now marked as completed. This is probably a separate issue from the one discussed here, but mentioning in case it is somewhat related.

**Comment by jmtaber129** [comment]
Okay, here's what I'd propose to resolve this:

Currently, we're explicitly removing `EstimateStage`s where the creator is the removed user. Instead, let's **simply leave these stages alone**. The `creatorUserId` doesn't seem to be used anywhere besides finding which stages to remove, so `EstimateStage`s aren't necessarily connected to a user within the UI.

Alternatively, we could update the `creatorUserId` of these would-be removed stages with the facilitator ID. I.e., the facilitator would take ownership of any orphaned estimate stages instead of just removing those stages entirely. However, this could prevent our audit ability if e.g. a customer wanted to know why an estimate stage exists when none of their team members created it.

**Comment by mattkrick** [comment]
tl;dr it was easier to delete than update.

the long answer:
- we use the `creatorUserId` to fetch the story from the integration for all team members (e.g. if I added the issue, when you come in the meeting, the meeting uses my credentials to fetch the story from jira for you. That way you can participate in the meeting without integrating with jira yourself)
-  if we change the `creatorUserId` to someone else, we'll need to first verify that they have a working integration that has permissions to view the story

To fix I'd propose we optimize for minimizing the number of different `creatorUserId`s:
- create a stack of estimate stages that are pending removal
- sort team members in some deterministic order (e.g. `teamMember.createdAt`)
- for each team member, for each estimate stage, fetch the issue. if it's a success, change the creatorUserId. if no team member can fetch the issue, remove the stage

**Comment by Dschoordsch** [comment]
I propose just assigning the facilitator or team lead. If either of them has the integration, assign the first match without verifying we can fetch the issue. It will work in most cases and feels less arbitrary than trying each team member.

**Comment by jmtaber129** [comment]
Ah, I think I understand now that I'm digging into integration tasks - it's not that we're using the _user ID_ to fetch the task from the integration, but rather the imported task may be associated with the user, so it's not as simple as changing the `creatorUserId`, but rather backfilling the would-be deleted task, too.

**Comment by mattkrick** [comment]
we currently don't have a UI for stale permissions though, if it fails the rest of the team will see "parabol can't connect to jira" which makes us look bad

**Comment by Dschoordsch** [comment]
On the other hand assigning random users will show the vote being posted by a random user, that seems not very trustworthy.

**Comment by mattkrick** [comment]
> On the other hand assigning random users will show the vote being posted by a random user, that seems not very trustworthy.

could you tell me more? i wasn't planning on involving votes in any way, this is just for setting `creatorUserId`

**Comment by Dschoordsch** [comment]
> * we use the `creatorUserId` to fetch the story from the integration for all team members (e.g. if I added the issue, when you come in the meeting, the meeting uses my credentials to fetch the story from jira for you. That way you can participate in the meeting without integrating with jira yourself)
> * if we change the `creatorUserId` to someone else, we'll need to first verify that they have a working integration that has permissions to view the story

Sorry, I was slightly mislead by your other comment
> we use the creatorUserId to fetch the story from the integration for all team members (e.g. if I added the issue, when you come in the meeting, the meeting uses my credentials to fetch the story from jira for you. That way you can participate in the meeting without integrating with jira yourself)
if we change the creatorUserId to someone else, we'll need to first verify that they have a working integration that has permissions to view the story

We don't use the `creatorUserId` to fetch tasks or set estimates, but instead we're using the `accessUserId` from the `TaskIntegration`. So we don't need to re-assign the `EstimateStage`, but we need to assign the `accessUserId` for the `Task` associated with that stage. We should not totally randomly assign that, as it will be used when a score is pushed to the integration and thus be visible in the integrated issue.

https://github.com/ParabolInc/parabol/blob/65d99cd3879adc86d0574d5165f0768f02e769f4/packages/server/graphql/types/EstimateStage.ts#L81
https://github.com/ParabolInc/parabol/blob/65d99cd3879adc86d0574d5165f0768f02e769f4/packages/server/graphql/mutations/setTaskEstimate.ts#L107-L108

**Comment by mattkrick** [comment]
ahh @Dschoordsch you're absolutely right! i forgot about moving access perms to the `accessUserId` 🤦 

@jmtaber129 it looks like you're right, we don't really use `creatorUserId` anywhere anymore.
- From a GDPR perspective, a document shared with the team is not PII, so we don't have to hard delete the stage when that user is hard deleted.
- it _is_ guaranteed to be non-null today. We could relax that constraint, or let the field go stale. since we don't have a `creatorUser` object on the estimate stage either would be perfectly OK

**Comment by mattkrick** [comment]
feels weird mutating the `creatorUserId`. the person who created the issue should be immutable. just because someone leaves a team doesn't mean we should change the truth of who created the stage.

chances are the `accessUserId` will be invalid when someone leaves a team, but that's a separate issue.  we should probably solve that in fetching an issue.  if `accessUserId` doesn't result in a successful fetch, we should try fetching it with someone else's credentials & update `accessUserId` upon success. this is because `accessUserId` could lose access to an issue outside of parabol, so it really operates as a hint, anyways.

**Comment by Dschoordsch** [comment]
I think we should not pick random members for accessing issues. It's fine if it's the creator of the task or the viewer, but imagine creator Cindy adds a task from project A, manager Mike sees that task and realizes that Cindy shouldn't have access and removes her from that team. The `accessUserId` now would not work anymore to fetch this task, but because manager Mike is still on the Parabol team, we would steal his credentials and fetch it.

**Comment by Dschoordsch** [comment]
If we don't want to delete the task, the only option left is improving the message we're showing. In an ideal world we would say something like:
"This Jira task was orphaned and has lost it's connection to Jira, press here to adopt it with your credentials"

**Comment by mattkrick** [comment]
gosh dang it you're right. we'll need some mutation like "claimTask" that sets the accessUserId to the person that clicks it.
